--- a/deps/v8/gypfiles/v8.gyp	2018-07-18 14:38:04.000000000 +0200
+++ b/deps/v8/gypfiles/v8.gyp	2018-07-28 08:06:26.705479200 +0200
@@ -1847,11 +1847,6 @@
             '../src/runtime/runtime-intl.cc',
           ],
         }],
-        ['OS=="win" and v8_enable_i18n_support==1', {
-          'dependencies': [
-            '<(icu_gyp_path):icudata',
-          ],
-        }],
       ],
     },
     {
